File: functions/fftspec.m
%
% FFTSPEC Fast-Fourier transform spectrum
%
%   spec = FFTSPEC(t,S)
%   Computes the magnitude FFT spectrum of the signal (S) on the time axis
%   (t).
%
%   [nu,spec] = FFTSPEC(t,S)
%   If two output arguments are requested, the frequency axis is returned
%   as well.
%
%   [nu,spec] = FFTSPEC(...,'Property',Value)
%   Additional (optional) arguments can be passed as property-value pairs.
%
% The properties to be passed as options can be set in any order.
%
%   'Type' - Type of spectrum to be returned ('real','imag','abs')
%
%   'ZeroFilling' - Number of elements in the output FFT spectrum
%                   (default = 2*length(S)).
%
%   'Apodization' - Use apodization window
%

% This file is a part of DeerLab. License is MIT (see LICENSE.md). 
% Copyright(c) 2019: Luis Fabregas, Stefan Stoll, Gunnar Jeschke and other contributors.


function varargout = fftspec(t,S,varargin)

if nargin<2
    error('Not enough inputs.');
end

%Parse optional input
[Type,ZeroFilling,useApodization] = parseoptional({'Type','ZeroFilling','Apodization'},varargin);

if isempty(Type)
    Type = 'abs';
else
    validateattributes(Type,{'char'},{},mfilename,'Type')
    allowedInput = {'abs','real','imag','complex'};
    validatestring(Type,allowedInput);
end

if isempty(useApodization)
    useApodization = true;
end
if isempty(ZeroFilling)
    ZeroFilling = 2*length(S);
else
    validateattributes(ZeroFilling,{'numeric'},{'scalar','nonnegative'},mfilename,'ZeroFilling')
end
validateattributes(useApodization,{'logical'},{'nonempty'},mfilename,'useApodization')
validateattributes(S,{'numeric'},{'2d','nonempty'},mfilename,'FitData')
validateattributes(t,{'numeric'},{'2d','nonempty','increasing'},mfilename,'t')
%Use column vectors
t = t(:);
S = S(:);

%If requested apply Hamming apodization window
if useApodization
    arg = linspace(0,pi,length(S));
    arg = arg.';
    ApoWindow = 0.54 + 0.46*cos(arg);
    S = S.*ApoWindow;
end

%Compute fft spectrum
Spectrum = fftshift(fft(S,ZeroFilling));

%Get the requested component/type of spectrum
switch Type
    case 'abs'
        Spectrum = abs(Spectrum);
    case 'real'
        Spectrum  = real(Spectrum);
    case 'imag'
        Spectrum = imag(Spectrum);
end

%If requested, get frequency axis and switch output order
if nargout>1
    FreqAxis = time2freq(t,ZeroFilling);
    varargout{1} = FreqAxis;
    varargout{2} = Spectrum;
else
    varargout{1} = Spectrum;
end


end

File: deerlab/fftspec.py

# fftspec.py - Fast-Fourier transform spectrum
# ---------------------------------------------------------------
# This file is a part of DeerLab. License is MIT (see LICENSE.md). 
# Copyright(c) 2019-2023: Luis Fabregas, Stefan Stoll and other contributors.

import numpy as np
from numpy.fft import fft, fftshift, fftfreq

def fftspec(V, t, mode='abs', zerofilling='auto', apodization=True):
    r"""
    Computes the Fast-Fourier Transform (FFT) spectrum of the input signal V.

    This function computes the FFT spectrum of the input signal ``V`` using the
    ``fft`` function from the ``numpy.fft`` module. The function allows to specify
    the type of spectrum to be returned, the zero-filling factor, and whether
    to apply a Hamming apodization window.
 
    Parameters
    ----------
    V : array_like
        Signal to be processed.
    
    t : array_like
        Time axis, in microseconds.
    
    mode : string, optional
        Type of spectrum to be returned (``'real'``,``'imag'``,``'abs'``), the default is ``'abs'``.
    
    zerofilling : scalar or ``'auto'``, optional
        Number of elements in the output FFT spectrum, the default is ``2*len(V)``.
    
    apodization : boolean, optional
        Use of a Hamming apodization window, the default is ``True``.

    Returns
    -------
    nu : ndarray
        Frequency axis, in megahertz
    spec : ndarray
        FFT spectrum.

    """
    
    if zerofilling == 'auto':
        zerofilling = 2*len(V)

    #If requested apply Hamming apodization window
    if apodization:
        arg = np.linspace(0,np.pi,len(V))
        ApoWindow = 0.54 + 0.46*np.cos(arg)
        V = V*ApoWindow

    #Compute fft spectrum
    spec = fftshift(fft(V,zerofilling))

    #Get the requested component/type of spectrum
    if mode == 'abs':
            spec = np.abs(spec)
    elif mode == 'real':
            spec  = spec.real
    elif mode == 'imag':
            spec = spec.imag
    else:
        raise KeyError("Invalid spectrum mode. Must be 'abs', 'real', or 'imag'. ")


    freq = fftshift(fftfreq(zerofilling,np.mean(np.diff(t))))

    return freq, spec 
